go/arrow/type_traits_float16.go

```diff
@@ -0,0 +1,95 @@
+package arrow
+
+import (
+	"encoding/binary"
+	"math"
+	"reflect"
+	"unsafe"
+)
+
+// Float16 traits
+
+type Float16 uint16
+
+// https://en.wikipedia.org/wiki/Half-precision_floating-point_format
+func NewFloat16(f float32) Float16 {
+	b := math.Float32bits(f)
+	sn := uint16((b >> 31) & 0x1)
+	exp := (b >> 23) & 0xff
+	res := int16(exp) - 127 + 15
+	fc := uint16(b >> 13) & 0x3ff
+	if exp == 0 {
```

Comment:
for multiple if-elses like this, Go style recommands a somewhat easier to read bool-switch:
```go
switch {
case exp == 0:
case exp == 0xff:
case res > 0x1e:
case res < 0x01:
}
```